### PR TITLE
Wordpress VIP

### PR DESCRIPTION
**Pull request by rgomezp**
Intention of this PR is to fix all blocking errors to make the plugin ready for submission to Wordpress VIP

### PHPCS
- Used PHP Code Sniffer to fix all errors
- `phpcs --standard=WordPressVIPMinimum -sp --basepath=. --ignore=vendor .`

### Other fixes
- Fixed issue on Gutenberg editor leading to double passes of the `send_notification_on_wp_post` hook after following [comment](https://github.com/WordPress/gutenberg/issues/15094#issuecomment-558986406)

---
This change is [Reviewable](https://reviewable.io/reviews/onesignal/onesignal-wordpress-plugin/219)

### CODE REVIEW

File: onesignal-public.php [hunk 1]
```diff
@@ -89,9 +35,11 @@ public static function onesignal_header()
         } ?>
     <?php
     if (defined('ONESIGNAL_DEBUG') && defined('ONESIGNAL_LOCAL')) {
-        echo '<script src="https://localhost:3001/sdks/OneSignalSDK.js" async></script>';
+        wp_register_script('local_sdk', plugins_url('https://localhost:3001/sdks/OneSignalSDK.js', __FILE__), array('jquery'), false, true);
+        wp_enqueue_script('local_sdk');
     } else {
-        echo '<script src="https://cdn.onesignal.com/sdks/OneSignalSDK.js" async></script>';
+        wp_register_script('remote_sdk', plugins_url('https://cdn.onesignal.com/sdks/OneSignalSDK.js', __FILE__), array('jquery'), false, true);
```
Comment (jkasten2) [inline onesignal-public.php]: We need to keep `async` in the `<script>` tag. Unfortunately `wp_register_script` does seem to support the async flag. Found a work around to still get it added even when using `wp_register_script` though.
https://stackoverflow.com/questions/18944027/how-do-i-defer-or-async-this-wordpress-javascript-snippet-to-load-lastly-for-fas

File: onesignal-public.php [hunk 2]
```diff
@@ -104,122 +52,118 @@ public static function onesignal_header()
...
-            if ($onesignal_wp_settings['prompt_action_message'] != '') {
-                echo "oneSignal_options['promptOptions']['actionMessage'] = '".OneSignalUtils::html_safe($onesignal_wp_settings['prompt_action_message'])."';\n";
+        if (array_key_exists('prompt_customize_enable', $onesignal_wp_settings) && $onesignal_wp_settings['prompt_customize_enable'] === '1') {
+            if (array_key_exists('prompt_action_message', $onesignal_wp_settings) && $onesignal_wp_settings['prompt_action_message'] !== '') {
+                echo "oneSignal_options['promptOptions']['actionMessage'] = '".esc_html($onesignal_wp_settings['prompt_action_message'])."';\n";
             }
-            if ($onesignal_wp_settings['prompt_example_notification_title_desktop'] != '') {
-                echo "oneSignal_options['promptOptions']['exampleNotificationTitleDesktop'] = '".OneSignalUtils::html_safe($onesignal_wp_settings['prompt_example_notification_title_desktop'])."';\n";
+            if (array_key_exists('prompt_example_notification_title_desktop', $onesignal_wp_settings) && $onesignal_wp_settings['prompt_example_notification_title_desktop'] !== '') {
```
Comment (jkasten2) [inline onesignal-public.php]: Can we make helper for the blank string and existence check? Checking the key twice is making the if statements long and harder to read and maintain.

**Comment by jkasten2** [comment]
We should setup Circle CI to run `phpcs` automatically since it will be important going forward that this is always passing. It doesn't have to be in this PR, but it should be in a follow up one.

File: onesignal-widget.php [hunk 3]
```diff
@@ -8,38 +8,38 @@ function __construct() {
...
+		$instance['title'] = ( ! empty( $new_instance['title'] ) ) ? wp_strip_all_tags( $new_instance['title'] ) : '';
+    	$instance['text'] = ( ! empty( $new_instance['text'] ) ) ? wp_strip_all_tags( $new_instance['text'] ) : '';
     
 		return $instance;
 	}
 
 	// Public display
 	function widget($args, $instance) {
-		echo $args['before_widget'];
+		echo esc_attr($args['before_widget']);
```
Comment (davisshaver) [inline onesignal-widget.php]: `wp_kses_post` here maybe?

File: onesignal-widget.php [hunk 4]
```diff
@@ -8,38 +8,38 @@ function __construct() {
...
-			echo $args['before_title'] . apply_filters( 'widget_title', $instance['title'] ). $args['after_title'];
+			echo esc_attr($args['before_title']) . esc_attr(apply_filters( 'widget_title', $instance['title'])). esc_attr($args['after_title']);
 		}
-    if ( ! empty( $instance['text'] ) ) {
-			echo '<a href="#" class="OneSignal-prompt">' . $instance['text'] . '</a>';
+    	if ( ! empty( $instance['text'] ) ) {
+			echo '<a href="#" class="OneSignal-prompt">' . esc_attr($instance['text']) . '</a>';
 		}
-		echo $args['after_widget'];
+		echo esc_attr($args['after_widget']);
```
Comment (davisshaver) [inline onesignal-widget.php]: Same as above

**Comment by davisshaver** [comment]
Excited to see VIP approval in progress! 🎉

File: onesignal-public.php (hunk 1 above)
Comment (rgomezp) [reply onesignal-public.php]: Great catch!

File: onesignal-admin.php [hunk 5]
```diff
@@ -827,21 +767,18 @@ public static function send_notification_on_wp_post($new_status, $old_status, $p
...
-    'timeout' => 60,
-    );
-
-                $response = wp_remote_post($onesignal_post_url, $request);
+                    'headers' => array(
+                                'content-type' => 'application/json;charset=utf-8',
+                                'Authorization' => 'Basic '.$onesignal_auth_key,
+                        ),
+                    'body' => wp_json_encode($fields),
+                    'timeout' => 3,
```
[the remark this answers is not in the document]
Comment (rgomezp) [reply onesignal-admin.php]: Unfortunately, PHPCS throws an error saying it is too long. What to do here?

File: onesignal-admin.php [hunk 6]
```diff
@@ -635,19 +609,18 @@ public static function send_notification_on_wp_post($new_status, $old_status, $p
 
             /* Settings related to creating a post involving the WordPress editor displaying the OneSignal meta box
              **********************************************************************************************************/
-
             /* Returns true if there is POST data */
-            $was_posted = !empty($_POST);
+            $was_posted = !empty(isset($_POST));
```
[the remark this answers is not in the document]
Comment (rgomezp) [reply onesignal-admin.php]: you're right. this is wrong

File: onesignal-admin.php [hunk 7]
```diff
@@ -543,25 +518,19 @@ public static function admin_custom_scripts()
         wp_enqueue_style('semantic-ui', plugin_dir_url(__FILE__).'views/css/semantic-ui.css', false, OneSignal_Admin::$RESOURCES_VERSION);
         wp_enqueue_style('site', plugin_dir_url(__FILE__).'views/css/site.css', false, OneSignal_Admin::$RESOURCES_VERSION);
 
-        wp_enqueue_script('jquery.min', plugin_dir_url(__FILE__).'views/javascript/jquery.min.js', false, OneSignal_Admin::$RESOURCES_VERSION);
-        wp_enqueue_script('semantic-ui', plugin_dir_url(__FILE__).'views/javascript/semantic-ui.js', false, OneSignal_Admin::$RESOURCES_VERSION);
-        wp_enqueue_script('jquery.cookie', plugin_dir_url(__FILE__).'views/javascript/jquery.cookie.js', false, OneSignal_Admin::$RESOURCES_VERSION);
-        wp_enqueue_script('site', plugin_dir_url(__FILE__).'views/javascript/site-admin.js', false, OneSignal_Admin::$RESOURCES_VERSION);
+        wp_enqueue_script('semantic-ui', plugin_dir_url(__FILE__).'views/javascript/semantic-ui.js', array('jquery'), OneSignal_Admin::$RESOURCES_VERSION);
```
[the remark this answers is not in the document]
Comment (rgomezp) [reply onesignal-admin.php]: This is a good question. I found this: https://digwp.com/2009/06/including-jquery-in-wordpress-the-right-way/

File: onesignal-admin.php [hunk 8]
```diff
@@ -487,7 +462,7 @@ public static function save_config_settings_form()
         if (array_key_exists('app_id', $_POST)) {
             // check_admin_referer dies if not valid; no if statement necessary
             check_admin_referer(OneSignal_Admin::$SAVE_CONFIG_NONCE_ACTION, OneSignal_Admin::$SAVE_CONFIG_NONCE_KEY);
-            $onesignal_wp_settings = OneSignal_Admin::save_config_page($_POST);
+            OneSignal_Admin::save_config_page($_POST);
```
[the remark this answers is not in the document]
Comment (rgomezp) [reply onesignal-admin.php]: done

**Comment by rgomezp** [comment]
> Excited to see VIP approval in progress! 🎉

Glad to see you're excited @davisshaver . Also, thanks for your suggestion. Please feel free to comment anywhere else you think we could make the plugin better. Cheers

File: onesignal-admin.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (jkasten2) [reply onesignal-admin.php]: @rgomezp Does provide any recommendations if a logger timeout is needed? Some linters provide a lint code to look up for more details too instead. Thinking they may suggest some sort of job if we need a longer timeout.

If they don't have any recommendations is there a way to ignore specific lints with a comment?

File: onesignal-admin.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (jkasten2) [reply onesignal-admin.php]: ah nice, included automatically if a script uses it!

File: onesignal-admin.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (rgomezp) [reply onesignal-admin.php]:
```
778 | ERROR   | Detected high remote request timeout. `timeout` is set to `60`. (WordPressVIPMinimum.Performance.RemoteRequestTimeout.timeout_timeout)
```

File: onesignal-admin.php [hunk 9]
```diff
@@ -580,6 +580,22 @@ public static function uuid($title)
         return substr($sha1, 0, 8).'-'.substr($sha1, 8, 4).'-'.substr($sha1, 12, 4).'-'.substr($sha1, 16, 4).'-'.substr($sha1, 20, 12);
     }
 
+    public static function exec_post_request($onesignal_post_url, $request, $retry_count) { 
+	if ($retry_count == 1) {
```
Comment (jkasten2) [inline onesignal-admin.php]: This should be `0` instead of `1`

File: onesignal-public.php [hunk 10]
```diff
@@ -75,11 +21,38 @@ public static function insert_onesignal_stamp()
     <?php
     }
 
+    function add_async_forescript($url)
```
[the remark this answers is not in the document]
Comment (jkasten2) [reply onesignal-public.php]: Still not fixed

File: onesignal-admin.php (hunk 9 above)
Comment (jkasten2) [reply onesignal-admin.php]: Commented on an older commit but the wrong value still applies here

File: onesignal-public.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (rgomezp) [reply onesignal-public.php]: Fixed

**Comment by rgomezp** [comment]
@jkasten2 made some changes to `onesignal-public.php`, sorry for the back and forth

**Comment by davisshaver** [comment]
@rgomezp  Has this been submitted to VIP for review yet? We may be asking to use it on a project and was curious if it's gone through process yet.

**Pull request merged**